### ISSUE TITLE
INDCCA / KyberKeyAgreement.java

### ISSUE THREAD

**Issue by asdfjkl** [fisherstevenk/kyberJCE#10]
only glanced at the code (and might misread it) but it looks to me that in 
`KyberKeyAgreement.java`, 
in the decryption, i.e. ` decrypt512` there is the following issue:

In case the FO-transform returns a failure (i.e. constant time comparing the arrays does not return 0), then instead of returning a pseudo-random value, you throw an exception. If so, this would very much destroy INDCCA?!

I.e. we should always return something, and in case of a failure a pseudo-random number (overwrite pre-k with z). E.g. see here https://github.com/pq-crystals/kyber/blob/master/ref/kem.c or in the spec...

Did I overlook something?

**Comment by fisherstevenk** [comment]
Well, we could return a psuedo-random value, but then the caller would have to verify if it was correct or not.  There's almost no point to the check at all at that point. 

C and Java are very different.  With an IllegalArgumentException, the caller immediately knows that something was wrong and they couldn't generate the proper shared secret.  If you can't decipher the cipher key into the correct shared secret, then it's of no use to you.

**Comment by asdfjkl** [comment]
If you raise an exception, that violates INDCCA of the protocol, which is the whole point of not using the INDCPA scheme directly, but applying the FO transform. Cf. Algorithm 9 of the official Kyber-Speficiation.

I think what they did in the Go implementation is to use the result of the constant time comparision (fail var) to overwrite pre-k with z (the whole part (fail & 0xFF ... completely vanishes if the comparison returns 0), which gives a time-constant implementation. So completely removing the if-check will probably fix this...

**Comment by fisherstevenk** [comment]
I understand what you're saying about the spec. but again, it's a difference of languages.

With the "Go" implementation, it can return 2 values, the ss and an error string.  If you walk through it, when the results are returned, the first thing it does is check the "err" variable for an issue and then runs the array compare.  But we can't do that in the real world, because we don't have the remote computed ss.

Their C implementation is the reference implementation with crypto_kem_dec being called from https://github.com/pq-crystals/kyber/blob/master/ref/PQCgenKAT_kem.c

Here's the relevant code:
       ```
         if ( (ret_val = crypto_kem_dec(ss1, ct, sk)) != 0) {
            printf("crypto_kem_dec returned <%d>\n", ret_val);
            return KAT_CRYPTO_FAILURE;
        }

        if ( memcmp(ss, ss1, CRYPTO_BYTES) ) {
            printf("crypto_kem_dec returned bad 'ss' value\n");
            return KAT_CRYPTO_FAILURE;
        }
```

In their impl, they hard coded a return of 0, hardly a proper way to indicate an error..  but they do check the result after calling the decrypt function.  They then do the array compare, but again, we don't have the remote ss so it's impossible to do this check.

This implementation of Kyber is for real world use, not just a "reference implementation".  If it was just for reference, I wouldn't have gone the route to create a full JCE out of it.

Somehow that error needs to get bubbled up to the user so they know that the data they provided is garbage.  I could implement a custom Exception instead of using an IllegalArgumentException and force the calling program to catch it.  That's about the only change I would make to this though.

**Comment by fisherstevenk** [comment]
Re-reading what you're saying..

For security of the protocol, the decrypt should just return whatever value gets generated, whether it's correct or not.  Is that what you're saying?

**Comment by asdfjkl** [comment]
Almost. In case of success, the correct value should be returned. If the comparison fails, than a pseudo-random key should be returned. See e.g. https://eprint.iacr.org/2017/634.pdf page 6 where they say

_"We stress that Kyber.Decaps never returns ⊥. Instead, in case re-encryption fails, it returns a pseudo-random key K := H(z, c), where z is a random, secret seed._

And last, to prevent timing side-channels, the running time should be constant, i.e. trying to avoid any if-check.

**Comment by fisherstevenk** [comment]
Gotcha.. I had to shift my head a little.  Thank you for your input.  I'll make the change when I get a chance.

**Comment by fisherstevenk** [comment]
Addressed in Pull #11
